File: Projects/codechef/UCL.java
package codechef;

import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;
import java.util.stream.Collectors;

class UCL {
    public static void main(String[] args) throws java.lang.Exception {
        Scanner stdin = new Scanner(System.in);
        int t = Integer.parseInt(stdin.nextLine());
        while (t-- > 0) {
            Map<String, Team> teams = new HashMap<>();
            for (int i = 0; i < 12; i++) {
                String[] tokens = stdin.nextLine().split(" ");
                if (!teams.containsKey(tokens[0]))
                    teams.put(tokens[0], new Team());
                if (!teams.containsKey(tokens[4]))
                    teams.put(tokens[4], new Team());
                int diff = Integer.parseInt(tokens[1]) - Integer.parseInt(tokens[3]);
                if (diff > 0)
                    teams.get(tokens[0]).plus(3, 0);
                else if (diff < 0)
                    teams.get(tokens[4]).plus(3, 0);
                else {
                    teams.get(tokens[0]).plus(1, 0);
                    teams.get(tokens[4]).plus(1, 0);
                }
                teams.get(tokens[0]).plus(0, diff);
                teams.get(tokens[4]).plus(0, -diff);
            }
            System.out.println(
                    teams.entrySet()
                            .stream()
                            .sorted(
                                    Map.Entry.comparingByValue(
                                            Comparator.comparingInt(Team::getPoint)
                                                    .thenComparingInt(Team::getGoal)
                                                    .reversed()
                                    )
                            )
                            .limit(2)
                            .map(Map.Entry::getKey)
                            .collect(Collectors.joining(" "))
            );
        }
    }

    private static class Team {
        private int point = 0;
        private int goal = 0;

        public void plus(int point, int goal) {
            this.point += point;
            this.goal += goal;
        }

        public int getPoint() {
            return this.point;
        }

        public int getGoal() {
            return this.goal;
        }
    }
}

File: Projects/codechef/UCL.kt
package codechef

import java.io.IOException
import java.util.*
import java.util.Map
import java.util.stream.Collectors

internal object UCL {
    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val stdin = Scanner(System.`in`)
        var t = stdin.nextLine().toInt()
        while (t-- > 0) {
            val teams: MutableMap<String, Team?> = HashMap()
            for (i in 0..11) {
                val tokens = stdin.nextLine().split(" ")
                if (!teams.containsKey(tokens[0])) {
                    teams[tokens[0]] = Team()
                }
                if (!teams.containsKey(tokens[4])) {
                    teams[tokens[4]] = Team()
                }
                val diff = tokens[1].toInt() - tokens[3].toInt()
                if (diff > 0) {
                    teams[tokens[0]]!!.plus(3, 0)
                } else if (diff < 0) {
                    teams[tokens[4]]!!.plus(3, 0)
                } else {
                    teams[tokens[0]]!!.plus(1, 0)
                    teams[tokens[4]]!!.plus(1, 0)
                }
                teams[tokens[0]]!!.plus(0, diff)
                teams[tokens[4]]!!.plus(0, -diff)
            }
            println(
                teams.entries
                .stream()
                .sorted(
                    Map.Entry.comparingByValue(
                        Comparator.comparingInt(Team::point)
                        .thenComparingInt(Team::goal)
                        .reversed()
                    )
                )
                .limit(2)
                .map<String> { it.key }
                .collect(Collectors.joining(" "))
            )
        }
        stdin.close()
    }

    private class Team {
        var point = 0
            private set
        var goal = 0
            private set

        fun plus(point: Int, goal: Int) {
            this.point += point
            this.goal += goal
        }

    }
}
